Automattic/jetpack
# PHPCS: Only Report on Changed Lines in pre-commit Hook

#### Changes proposed in this Pull Request:
1. Look at all staged files. Calculate their changed lines (new or changed, not deleted: anything in the new version of the file that was not in the old version of the file).
2. Run PHPCS on those files.
3. From the PHPCS report, filter out all errors that are on lines not included in the list from step 1.

This PR does not get rid of the whitelist. Before we can get rid of the whitelist, we should at least run the autofixer on everything (yay syncing those changes to WP.com!).

I've added `3rd-party/` to the whitelist, though, to make it easier to test some things.

#### Testing instructions:
Before you do anything else:
1. Check out this branch: `git checkout update/phpcs-changed-lines-only`
2. `yarn install`
3. `composer install`

##### 3rd-party/vaultpress.php
This file has just a few errors and none are automatically fixable.

1. `vendor/bin/phpcs 3rd-party/vaultpress.php`
2. See several errors. Note that one of them is on line 20.
3. Edit 3rd-party/vaultpress.php. Change line 20, but don't fix the PHPCS error:
```diff
--- a/3rd-party/vaultpress.php
+++ b/3rd-party/vaultpress.php
@@ -17,7 +17,7 @@ function jetpack_vaultpress_rewind_enabled_notice() {
 	?>
 	<div class="notice notice-success vp-deactivated">
 		<h2 style="margin-bottom: 0.25em;"><?php _e( 'Jetpack is now handling your backups.', 'jetpack' ); ?></h2>
-		<p><?php _e( 'VaultPress is no longer needed and has been deactivated.', 'jetpack' ); ?></p>
+		<p><?php _e( 'VaultPress is no longer needed and has been deactivated!', 'jetpack' ); ?></p>
 	</div>
 	<style>#vp-notice{display:none;}</style>
 	<?php
```
4. `git add 3rd-party/vaultpress.php`
5. `git commit`
6. See the commit blocked, but only by the one error!
7. Fix the error:
```diff
--- a/3rd-party/vaultpress.php
+++ b/3rd-party/vaultpress.php
@@ -17,7 +17,7 @@ function jetpack_vaultpress_rewind_enabled_notice() {
 	?>
 	<div class="notice notice-success vp-deactivated">
 		<h2 style="margin-bottom: 0.25em;"><?php _e( 'Jetpack is now handling your backups.', 'jetpack' ); ?></h2>
-		<p><?php _e( 'VaultPress is no longer needed and has been deactivated!', 'jetpack' ); ?></p>
+		<p><?php esc_html_e( 'VaultPress is no longer needed and has been deactivated!', 'jetpack' ); ?></p>
 	</div>
 	<style>#vp-notice{display:none;}</style>
 	<?php

```
8. `git add 3rd-party/vaultpress.php`
9. `git commit`
10. See the pre-commit pass, even though you didn't fix all the other errors in the file.
11. (Cancel the commit by exiting the editor without saving the commit message file.)
12. (`git reset HEAD 3rd-party/vaultpress.php; git checkout -- 3rd-party/vaultpress.php`)

**Note**: If we had not doen step 8 (`git add`), the pre-commit check would *still have passed* since PHPCS tests what's in the working copy, not what's staged. That is, you can "trick" PHPCS into passing the commit without actually including any error fixes. This has not changed in this PR, but is something we should discuss.

##### 3rd-party/woocommerce.php
This file has just a few errors, and some of them are automatically fixable.

1. `vendor/bin/phpcs 3rd-party/woocommerce.php`
2. See several errors. Note that one of them is on line 97.
3. Edit 3rd-party/woocommerce.php. Change line 97, but don't fix the PHPCS error:
```diff
--- a/3rd-party/woocommerce.php
+++ b/3rd-party/woocommerce.php
@@ -94,7 +94,7 @@ function jetpack_woocommerce_infinite_scroll_style() {
 	wp_add_inline_style( 'woocommerce-layout', $custom_css );
 }
 
-function jetpack_woocommerce_lazy_images_compat() {
+function jetpack_woocommerce_lazy_images_compat( $foo = false ) {
 	wp_add_inline_script( 'wc-cart-fragments', "
 		jQuery( 'body' ).bind( 'wc_fragments_refreshed', function() {
 			jQuery( 'body' ).trigger( 'jetpack-lazy-images-load' );
```
4. `git add 3rd-party/woocommerce.php`
5. `git commit`
6. See the commit blocked, but only by the one error!
7. See also that "PHPCS detected and fixed some automatically fixable errors".
8. `git diff --staged` to see that those automatic fixed really have been applied.
9. Fix the error:
```diff
--- a/3rd-party/woocommerce.php
+++ b/3rd-party/woocommerce.php
@@ -94,6 +94,11 @@ function jetpack_woocommerce_infinite_scroll_style() {
 	wp_add_inline_style( 'woocommerce-layout', $custom_css );
 }
 
+/**
+ * OK
+ *
+ * @param bool $foo bar.
+ */
 function jetpack_woocommerce_lazy_images_compat( $foo = false ) {
 	wp_add_inline_script(
 		'wc-cart-fragments',
```
10. `git add 3rd-party/woocommerce.php`
11. `git commit`
12. See the pre-commit pass, even though you didn't fix all the other errors in the file.
13. (Cancel the commit by exiting the editor without saving the commit message file.)
14. Also see that there were some non-blocking warnings in the changed lines.
15. (`git reset HEAD 3rd-party/woocommerce.php; git checkout -- 3rd-party/woocommerce.php`)

#### TODO
* [ ] Does any of this work on Windows?
* [x] Fix staged/working copy inconsistencies (fixed in replacement PR: #11545).
* [x] Provide easy (non-`git commit`) way of running these line filtered sniffs (fixed in replacement PR: #11545).

#### Proposed changelog entry for your changes:
None required.
